src/v/cluster/controller_backend.cc

```diff
@@ -1009,27 +1030,22 @@ controller_backend::process_partition_reconfiguration(
 }
 
 bool controller_backend::can_finish_update(
+  uint64_t current_retry,
   topic_table_delta::op_type update_type,
-  const std::vector<model::broker_shard>& current_replicas,
-  const std::vector<model::broker_shard>& previous_replicas) {
+  const std::vector<model::broker_shard>& current_replicas) {
     // force abort update may be finished by any node
     if (update_type == topic_table_delta::op_type::force_abort_update) {
         return true;
     }
-    // update may be finished by a node that was added to replica set
-    if (!has_local_replicas(_self, previous_replicas)) {
-        return true;
-    }
-    // finally if no nodes were added to replica set one of the current replicas
-    // may finish an update
-
-    auto added_nodes = subtract_replica_sets(
-      current_replicas, previous_replicas);
-    if (added_nodes.empty()) {
-        return has_local_replicas(_self, current_replicas);
-    }
+    /**
+     * Use retry count to determine which node is eligible to dispatch update
+     * finished. Using modulo division allow us to round robin between
+     * candidates
+     */
+    const model::broker_shard& candidate
+      = current_replicas[current_retry % current_replicas.size()];
```

Comment:
This feels a bit artificial to me - round robins are not synchronized across replicas, so probably not a strong defense against duplicates (though of course they are relatively harmless).

BTW I forgot - why don't we just let the partition leader dispatch the update_finished command? Seems like a more natural choice.

Reply:
actually, we can do it on leader

Reply:
very good point

Reply:
actually the leadership approach will not work when upgrading from previous version (leader may be on the old node).  For now i will leave version with the modulo division based approach. in the next PR of this series i introduce a feature so then i am going to fix it